### ISSUE TITLE
integration with major backup vendors software

### ISSUE THREAD

**Issue by alexeyshishkin01** [pgbackrest/pgbackrest#1394]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version:

2. PostgreSQL version:

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:

4. Did you install pgBackRest from source or from a package?

5. Please attach the following as applicable:
    - `pgbackrest.conf` file(s)
    - `postgresql.conf` settings applicable to pgBackRest (`archive_command`, `archive_mode`, `listen_addresses`, `max_wal_senders`, `wal_level`, `port`)
    - errors in the postgresql log file before or during the time you experienced the issue
    - log file in `/var/log/pgbackrest` for the commands run (e.g. `/var/log/pgbackrest/mystanza_backup.log`)

7. Describe the issue:

This question is about backup of Postgres the way Oracle RMAN does.
Oracle RMAN integrates with major backup vendors software (Commvault, Dell/EMC, IBM, Veritas etc.) to backup directly to tape/VTL/PBBA etc.
Most of enterprise customers use sophisticated backup infrastructure:
- backup hardware
- backup network (SAN or LAN)
- backup software

For Postgres it's only possible to backup to disk.
That's why enterprise customers have to allocate at least temporary disk for backup.
In case of multi-terabyte database it's really expensive even for deep-pocket customers.

**Comment by dwsteele** [comment]
Up until recently this feature would not have been very useful, since at least a few backups and their WAL should be stored locally for replication, rebuilding standbys, etc.

Now that we can have multiple repositories, it might make sense to have one of them on tape/VTL/PBBA. However, the trend these days is to use objects storage (e.g. S3) for offsite backup. It is not clear that working with tape/VTL/PBBA storage is worth the effort and it has not been a big ask.

One solution I have seen in the field is to use the `archive-copy` option once a week to create a self-contained backup and then point the backup software at that backup directory or copy the backup directory elsewere.

Furthermore, all the tape/VTL/PBBA software you have named here is proprietary closed source. We don't have any licensing for it and it was be hard/impossible to integration with our unit/integration tests.

In short, this is not a feature we are currently planning.

If you look in `src/storage` you'll find several examples of storage drivers so you are more than welcome to create your own tape driver if the existing interface permits it.

**Comment by dwsteele** [comment]
Also, objects stores backed by PBBAs (on of off prem) are becoming quite common.

**Comment by alexeyshishkin01** [comment]
@dwsteele , thanks for the answers !

some thoughts/questions from my side:

'Furthermore, all the tape/VTL/PBBA software you have named here is proprietary closed source.'
- these proprietary closed source backup software vendors have lion share of backup software market and most of the enterprise customers use such software

'It is not clear that working with tape/VTL/PBBA storage is worth the effort and it has not been a big ask.'
- I work with large enterprise customers who's spent a lot of money on their backup infra for MSFT/ORCL and would like to reuse it for Postgres as well

it looks like Postgres is not used by enterprise customers in your region so far, is it ?

**Comment by dwsteele** [comment]
> it looks like Postgres is not used by enterprise customers in your region so far, is it ?

So far the large enterprises we have worked with are using one of these three methods:

1) Backup the entire repo to tape
2) Backup the last full backup (with WAL included by `archive-copy`) to tape
3) Backup to on-prem cloud storage on a backup appliance

We were considering adding tape support about four years ago for a very large organization but they ended up implementing #3 and it was not necessary.

We know from experience that backing up to tape as the only backup method is generally not viable. WAL is needed to feed replicas, backups are needs to rebuild replicas on failover, etc. So some on-disk presence is generally a must, and that being the case traditional backup software works fine against the repo.

It seems what you are asking for is the ability to have the *only* repo be on tape and that appears to be too narrow a use case to be worth the considerable effort it would take to implement it. We have limited development resources and must pick our battles.

**Comment by alexeyshishkin01** [comment]
@dwsteele , thanks for your answers !

when you say 'backup the entire repo to tape' what do you mean :
- what is 'repo' ? - no guess from my side
- what is 'tape' ? - is it single-drive device or multi-drive device / tape library with robot ? is it real tape device or virtual tape library ?
- do you support particular type of tape vendor or any type tapes ?

'It seems what you are asking for is the ability to have the only repo be on tape and that appears to be too narrow a use case to be worth the considerable effort it would take to implement it.'
my question is still the same:
- enterprise customers use backup software and backup hardware to backup their databases (IBM/MSFT/ORCL) over SAN/LAN on-premise
- they integrate database native backup tools (like Oracle RMAN) with backup software (Commvault, Micro Focus, Veritas) in order to backup to almost any kind of backup hardware (disk, tape, VTL, PBBA)
- many of enterprise customers prefer PBBA (EMC Data Domain) because of its price per gigabyte, performance and advanced features (deduplication, compression)
- my idea is either to integrate with backup software vendors via Postgres native backup tool or build/find new backup software (open/closed source) to backup huge Postgres databases the similar way to any backup hardware
- it might sound too ambitious or/and unrealistic but it looks like enterprise customers don't leave us another option :-)

**Comment by pgstef** [comment]
Hi,


>     * what is 'repo' ? - no guess from my side

The [user-guide](https://pgbackrest.org/user-guide.html#quickstart/create-repository) is pretty clear about this one: _The repository is where pgBackRest stores backups and archives WAL segments._

The repository is one of the main concept of this tool. Its configuration/location and what is stored there is the base you need to understand to figure out why you couldn't stream WAL archives to tapes (or to other tools) one by one. And WAL archives are the root of PG point-in-time recovery.

> my question is still the same:
>     * enterprise customers use backup software and backup hardware to backup their databases (IBM/MSFT/ORCL) over SAN/LAN on-premise
>     * they integrate database native backup tools (like Oracle RMAN) with backup software (Commvault, Micro Focus, Veritas) in order to backup to almost any kind of backup hardware (disk, tape, VTL, PBBA)
>     * many of enterprise customers prefer PBBA (EMC Data Domain) because of its price per gigabyte, performance and advanced features (deduplication, compression)
>     * my idea is either to integrate with backup software vendors via Postgres native backup tool or build/find new backup software (open/closed source) to backup huge Postgres databases the similar way to any backup hardware

IMHO, or you get a "generic" tool that does everything (but not completely/correctly) or you get a "specific" tool that does support all the features of 1 technology. And that's the choice here.

As far as I know, today, there's no PG backup tool supporting what you describe. David was pretty clear. This project is open-source, with limited resources. How could we integrate "closed-source/proprietary" softwares in our CI (continuous integration - which is a part of this project quality check) tests without them? I guess we'd be open to contributions (respecting the contributing guide) though...

Furthermore, you're talking about "deduplication" and "compression". That's already something supported in most PG backup tools (using incremental backups,...). And once again, what's the point of "deduplicating" WAL archives which are supposed to be unique? What you describe seems a bit off the PG PITR reality...

>     * it might sound too ambitious or/and unrealistic but it looks like enterprise customers don't leave us another option :-)

Well perhaps "your" enterprise customers don't leave you the choice. I'm pretty sure the companies David and I are working for also have big enterprise customers... Your assumption seems a bit rude here.


Kind Regards,

**Comment by alexeyshishkin01** [comment]
@pgstef , thanks for your comments !

may I ask you to answer this question as well:

"
what is 'tape' ? - is it single-drive device or multi-drive device / tape library with robot ? is it real tape device or virtual tape library ?
"

many thanks !

**Comment by dwsteele** [comment]
> what is 'tape' ?

In this case I am using "tape" to denote the method of backup/retrieval, i.e. that it must be serialized, as was common last time I was looking at this area. But I'm not familiar with the APIs for the products you are naming, so I could be off-base here. If they are not "tape-like" then they can provide an S3 interface, as is common now.

Enterprise backup solutions that support the S3 API in either cloud or appliance form are quite common now: IBM Cloud Storage, Dell appliances, Veeam, Backblaze, not to mention AWS, etc.

**Comment by alexeyshishkin01** [comment]
thanks, @dwsteele !

ok, we have to devices: 'disk' and 'tape', the latter one is with serialized access

what Oracle did was really nice - they created API for their native backup tool 'RMAN - Recovery MANager' which backup software vendors have to use in order to backup Oracle database via RMAN to backup devices

essentially, backup software vendors job was to provide access/path to backup hardware devices and Oracle RMAN did the backup/restore job itself

so, backup software vendors have to follow Oracle, not vice versa 

since Oracle was quite popular at that time, backup software vendors did follow Oracle and conform to Oracle's API

today Postgres popularity is comparable to Oracle's one, so one can create similar API and push it to Postgres community
